python/lsst/summit/utils/astrometry/fastStarTrackerAnalysis.py

```diff
@@ -0,0 +1,551 @@
+        for itemName in self.__slots__:
+            v = getattr(self, itemName)
+            if isinstance(v, int):  # print ints as ints
+                retStr += f'{itemName} = {v}\n'
+            elif isinstance(v, float):  # but round floats at 3dp
+                retStr += f'{itemName} = {v:.3f}\n'
+            elif itemName == 'moments':  # and don't spam the full moments
+                retStr += f'moments = {type(v)}\n'
+            elif itemName == 'bbox':  # and don't spam the full moments
+                retStr += f'bbox = lsst.geom.{repr(v)}\n'
+            elif itemName == 'cutout':  # and don't spam the full moments
+                if v is None:
+                    retStr += 'cutout = None\n'
+                else:
+                    retStr += f'cutout = {type(v)}\n'
+        return retStr
+
+
+def analyzeFastStarTrackerImage(filename, boxSize, attachCutouts=True):
+    """Analyze a single FastStarTracker image.
+
+    Parameters
+    ----------
+    filename : `str`
+        The full
+    boxSize : `int`
+        The size of the box to put around each source for measurement.
+    attachCutouts : `bool`, optional
+        Attached the cutouts to the ``Source`` objects? Useful for
+        debug/plotting but adds memory usage.
+
+    Returns
+    -------
+    sources : `list` of
+              `lsst.summit.utils.astrometry.fastStarTrackerAnalysis.Source`
+        The sources in the image, sorted by rawFlux.
+    """
+    exp = tifToExp(filename)
+    footprintSet = detectObjectsInExp(exp)
+    footprints = footprintSet.getFootprints()
+    bgMean, bgStd = getBackgroundLevel(exp)
+
+    sources = []
+    for footprint in footprints:
+        source = Source()
+        source.nSourcesInImage = len(footprints)
+        source.parentImageWidth, source.parentImageHeight = exp.getDimensions()
+
+        centroid = footprint.getCentroid()
+        bbox = getBboxAround(centroid, boxSize, exp)
+        source.bbox = bbox
+        cutout = exp.image[bbox].array
+        if attachCutouts:
+            source.cutout = cutout
+        source.centroidX = centroid[0]
+        source.centroidY = centroid[1]
+        source.rawFlux = getFlux(cutout, bgMean)
+        source.imageBackground = bgMean
+        source.imageStddev = bgStd
+        source.nPix = countOverThresholdPixels(cutout, bgMean, bgStd)
+
+        moments = galsim.hsm.FindAdaptiveMom(galsim.Image(cutout))
+        source.moments = moments
+        source.hsmFittedFlux = moments.moments_amp
+        source.hsmCentroidX = moments.moments_centroid.x + bbox.minX - 1
+        source.hsmCentroidY = moments.moments_centroid.y + bbox.minY - 1
+        source.localCentroidX = moments.moments_centroid.x - 1
+        source.localCentroidY = moments.moments_centroid.y - 1
+        sources.append(source)
+    return sortSourcesByFlux(sources)
+
+
+def checkResultConsistency(results, maxAllowableShift=5, silent=False):
+    """Check if a set of results are self-consistent.
+
```

Comment:
How likely is it that you will have issues with sources at the edge of detectors? Having the brightest star at an edge seems like it will have a lot of unwanted consequences if it scatters or drifts off the detector.

Reply:
Quite likely, and I totally agree that will cause problems. That was kind of the thrust of the consistency checking helper function. My thinking is that if it slips off the edge the you either get a massive jump in centroid, or a different number of source counts, or a huge jump in brightness, or something else. Hopefully that will help people find things when it's happening, but I don't think there's much that can be done about it, at this stage at least.

Reply:
If it looks necessary, I think that coming back to all this code and implementing a better way of doing source association between the images would be a good plan, but 1) this was all written on a plane, and 2) all the data so far has no problems in source association, so I just didn't need to do it, and so didn't given the great urgency with all this stuff 🙂